### PR TITLE
Add NationCred smart contract (#24)

### LINKED ISSUE #24: Add NationCred smart contract
Create smart contract for storing the passport IDs of active Nation3 Citizens.

### PR DESCRIPTION
**Pull request by aahna-ashina**
## Related GitHub Issue

closes #24

## How Has This Been Tested?

- [x] `npx hardhat test`
- [x] Status checks pass
- [x] Works on Sepolia:  https://sepolia.etherscan.io/tx/0x72d669c18cb4a02db1f680c6bc123e8d60f43834207e672dae682ea788d55830
  - Copied an array of active passportIDs from https://github.com/nation3/nationcred-datasets/blob/main/nationcred/output/nationcred-active-citizens.csv
  - Screenshot 2023-03-23 at 4 31 15 PM
- [x] Deployed to Sepolia after refactoring: 
 https://sepolia.etherscan.io/tx/0xb68bd231719c951da20e6368d290addefbff2d171ef65b8dcefd25d5d3459489
  - 75,148 gas for for setting 5 active citizens:  https://sepolia.etherscan.io/tx/0x56752402c1f562083a5e131a92e126bd2a9b82e654793b87a50c6191db3c5307
  - 41,524 gas for setting 8 active citizens:  https://sepolia.etherscan.io/tx/0xfb7d1c9e55ce791f9dbd1ef6385a40e2046fafa5908421baf6d2cf0719f4e116
  - 47,030 gas for setting 25 active citizens:  https://sepolia.etherscan.io/tx/0x363678590109374b60ccad198d857f017b5d8bf16fabc3ca6490c9a5c1515101
- [x] Works on Mainnet

Code coverage:

```
------------------|----------|----------|----------|----------|----------------|
File              |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
------------------|----------|----------|----------|----------|----------------|
 contracts/       |    92.86 |       25 |    92.31 |    93.75 |                |
  Discord.sol     |      100 |      100 |      100 |      100 |                |
  Discourse.sol   |      100 |      100 |      100 |      100 |                |
  GitHub.sol      |      100 |      100 |      100 |      100 |                |
  Greeter.sol     |      100 |      100 |      100 |      100 |                |
  INationCred.sol |      100 |      100 |      100 |      100 |                |
  NationCred.sol  |       80 |       25 |       75 |       80 |          20,21 |
------------------|----------|----------|----------|----------|----------------|
All files         |    92.86 |       25 |    92.31 |    93.75 |                |
------------------|----------|----------|----------|----------|----------------|
```

## Are There Admin Tasks?

- [x] Deploy to Sepolia: 
 https://sepolia.etherscan.io/tx/0xb68bd231719c951da20e6368d290addefbff2d171ef65b8dcefd25d5d3459489
- [x] Deploy to Mainnet:  https://etherscan.io/tx/0xda59584d17e8d26f14666bbcf228d03b53565712b2d2e12bfee490a0a1383590

### CODE REVIEW

File: contracts/NationCred.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    constructor() {
+        owner = address(msg.sender);
+    }
+
+    function setOwner(address owner_) public {
+        require(msg.sender == owner, "Only the owner can call this function");
+        owner = owner_;
+    }
+
+    function setActiveCitizens(uint16[] calldata passportIDs_) public {
```
Comment (johnmark13) [inline contracts/NationCred.sol:24]: It took me a moment to see that the parameter name was different ad I was scratching my head as to what this function did, maybe just call it `activePassportIDs` or updated, or something to differentiate the variables more clearly.

File: test/ NationCred.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    await nationCred.setActiveCitizens([40, 42, 44]);
+
+    await nationCred.setActiveCitizens([42]);
+
+    expect(await nationCred.isActive(40)).to.equal(false);
+    expect(await nationCred.isActive(42)).to.equal(true);
+    expect(await nationCred.isActive(44)).to.equal(false);
+  });
+});
```
Comment (johnmark13) [inline test/ NationCred.ts:49]: ❤️

**Review by johnmark13 — CHANGES_REQUESTED** [review]

File: contracts/NationCred.sol (hunk 1 above)
Comment (aahna-ashina) [reply contracts/NationCred.sol:24]: Good point, will fix! 👍

Comment (aahna-ashina) [reply contracts/NationCred.sol]: Fixed:  https://github.com/nation3/nationcred-contracts/pull/45/commits/3e4c45958c89d34f0db0ff4d663e2d1b4c07404b

**Comment by aahna-ashina** [comment]
https://github.com/nation3/nationcred-contracts/pull/45/commits/17e231f82f4af9d393c5f62c9f8fdfdf135687fd Reduced gas cost:

Before:
```

|  NationCred  ·  setActiveCitizens  ·      47049  ·      143286  ·     111207  ·            3  ·          -  │
```


After:
```
|  NationCred  ·  setActiveCitizens  ·       40180  ·      74764  ·      63236  ·            3  ·          -  │
```

**Review by johnmark13 — APPROVED** [review]

**Pull request merged**